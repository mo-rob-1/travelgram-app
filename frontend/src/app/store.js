import { configureStore } from "@reduxjs/toolkit";
import authReducer from "../features/auth/authSlice";
import imageReducer from "../features/images/imageSlice";

export const store = configureStore({
  reducer: {
    auth: authReducer,
    images: imageReducer,
  },
});
