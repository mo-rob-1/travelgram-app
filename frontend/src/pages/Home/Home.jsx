import { useEffect, useState } from "react";
import axios from "axios";
import { Section, List, Figure, Location, Username, IconWrapper, Title, LoginLink, Span } from "./Home.styled";
import ModalImage from "react-modal-image";
import { GoLocation } from "react-icons/go";
import { AiOutlineUser } from "react-icons/ai";
import { useSelector } from "react-redux";

function Home() {
  const [images, setImages] = useState([]);
  const [limitImages, setLimitImages] = useState([]);
  const [loading, setLoading] = useState(true);

  const { user } = useSelector((state) => state.auth);

  useEffect(() => {
    axios
      .get("/api/images")
      .then((res) => {
        setImages(res.data.reverse());
        setLoading(false);
      })
      .catch((err) => {
        console.log(err);
      });
  }, []);

  useEffect(() => {
    if (images.length > 0) {
      const limit = images.slice(0, 9);
      setLimitImages(limit);
    }
  }, [images]);

  if (loading) {
    return (
      <Section>
        <h2>Loading...</h2>
      </Section>
    );
  }

  return (
    <Section>
      {images.length === 0 && (
        <div>
          <Title>All Images</Title>
          <h2>Login or register to be the first person to add an image</h2>
        </div>
      )}
      {user ? (
        <div>
          <Title>All Images</Title>
          <List>
            {images.map((image) => (
              <li key={image.cloudinary_id}>
                <Figure>
                  <div>
                    <ModalImage
                      small={image.image}
                      large={image.image}
                      alt={image.caption}
                      className="modalImg"
                      data-testid="image"
                    />
                  </div>
                  <Username>
                    <IconWrapper>
                      <AiOutlineUser style={{ marginRight: "4px" }} /> {image.username}
                    </IconWrapper>
                  </Username>
                  <Location>
                    <IconWrapper>
                      <GoLocation style={{ marginRight: "4px" }} /> {image.imageLocation}
                    </IconWrapper>
                  </Location>
                </Figure>
              </li>
            ))}
          </List>
        </div>
      ) : (
        <div>
          <Title>
            Welcome to <Span>travel</Span>Gram
          </Title>
          <List>
            {limitImages.map((image) => (
              <li key={image.cloudinary_id}>
                <Figure>
                  <div>
                    <ModalImage
                      small={image.image}
                      large={image.image}
                      alt={image.caption}
                      className="modalImg"
                      data-testid="image"
                    />
                  </div>
                  <Username>
                    <IconWrapper>
                      <AiOutlineUser style={{ marginRight: "4px" }} /> {image.username}
                    </IconWrapper>
                  </Username>
                  <Location>
                    <IconWrapper>
                      <GoLocation style={{ marginRight: "4px" }} /> {image.imageLocation}
                    </IconWrapper>
                  </Location>
                </Figure>
              </li>
            ))}
          </List>
          <div style={{ textAlign: "center", marginTop: "4rem", marginBottom: "4rem" }}>
            <LoginLink to="/login">Login to view all images</LoginLink>
          </div>
        </div>
      )}
    </Section>
  );
}

export default Home;
