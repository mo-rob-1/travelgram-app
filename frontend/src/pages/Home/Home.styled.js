import styled from "styled-components";
import { Link } from "react-router-dom";

export const List = styled.ul`
  list-style-type: none;
  display: grid;
  grid-gap: 1rem;
  @media (min-width: 768px) {
    grid-template-columns: repeat(2, 1fr);
    grid-template-rows: 1fr;
    margin-top: 2rem;
  }
  @media (min-width: 1024px) {
    grid-template-columns: repeat(3, 1fr);
    gap: 1.5rem;
  }
`;

export const Figure = styled.figure`
  position: relative;
`;

export const Location = styled.figcaption`
  position: absolute;
  bottom: 0;
  padding: 1rem;
  color: #fff;
  background-color: #222222b8;
  font-size: 14px;
`;

export const Username = styled.figcaption`
  position: absolute;
  top: 0;
  right: 0;
  background: #222222b8;
  color: #fff;
  padding: 1rem;
  font-size: 14px;
`;

export const IconWrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const Section = styled.section`
  margin-top: 1rem;

  @media (min-width: 1024px) {
    margin-top: 3rem;
  }
`;

export const Title = styled.h1`
  margin-top: 0;
  @media (min-width: 768px) {
    margin-top: 1rem;
  }
`;

export const LoginLink = styled(Link)`
  text-align: center;
  color: #fff;
  padding: 1rem;
  background-color: #222;
  text-decoration: none;
`;

export const Span = styled.span`
  font-weight: 400;
`;
