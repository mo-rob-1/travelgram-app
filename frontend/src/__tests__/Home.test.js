import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { Provider } from "react-redux";
import { store } from "../app/store";
import { BrowserRouter as Router } from "react-router-dom";
import Home from "../pages/Home/Home";
import axios from "axios";

describe("Home", () => {
  it("renders the homepage", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Provider store={store} history={history}>
        <Router>
          <Home />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the images", () => {
    const { findByTestId } = render(
      <Provider store={store}>
        <Router>
          <Home />
        </Router>
      </Provider>
    );

    findByTestId(() => {
      // get by data-testid
      const image = document.querySelector("[data-testid='image']");
      expect(image).toBeInTheDocument();
    });
  });

  it("checks for axios error", async () => {
    await axios
      .get("http://localhost:5000/api/images")
      .then((res) => {
        expect(res.status).toBe(200);
      })
      .catch((err) => {
        expect(err).toBeTruthy();
      });
  });

  it("checks for axios success", async () => {
    await axios
      .get("http://localhost:5000/api/images")
      .then((res) => {
        expect(res.status).toBe(200);
      })
      .catch((err) => {
        expect(err).toBeFalsy();
      });
  });
});
