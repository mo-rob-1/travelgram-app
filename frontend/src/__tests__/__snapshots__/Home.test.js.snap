// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders the homepage 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Welcome to travelGram
    </h1>
    <ul />
  </div>
</DocumentFragment>
`;
